### ISSUE TITLE
Inaccurate Solutions for VRP using VQE and VehicleRouting classes

### ISSUE THREAD

**Issue by RedWyrm13** [qiskit-community/qiskit-optimization#636]
### Environment

- **Qiskit Optimization version**: 0.6.1
- **Python version**: Python 3.11
- **Operating system**: Windows 11 Enterprise

### What is happening?

When using the [Vehicle Routing](https://github.com/qiskit-community/qiskit-optimization/blob/main/qiskit_optimization/applications/vehicle_routing.py) class, I am getting solutions that are violating the constraints and not matching the CPLEX solution from the classical optimizer in [this](https://github.com/qiskit-community/qiskit-optimization/blob/main/docs/tutorials/07_examples_vehicle_routing.ipynb) tutorial.

### How can we reproduce the issue?

```


from qiskit.primitives import Estimator as PrimEstimator, StatevectorSampler as PrimSampler
from qiskit.circuit.library import EfficientSU2, RealAmplitudes
from qiskit_algorithms.optimizers import SPSA, COBYLA
from qiskit_optimization.applications import VehicleRouting
from qiskit_optimization.converters import QuadraticProgramToQubo
from qiskit_algorithms import VQE
import numpy as np

###These are imported functions###########################################
import numpy as np

#This function takes in a list of strings in the format ['000', '001', 010', '100'] and returns the path that it represents
def format_solution(entire_bitstring):
    
    slices = get_slices(entire_bitstring)

    path = []
    #This loops over the outter list. i 
    for i, bitstring in enumerate(slices):
        edge1 = i
        
        #This loops over each bitstring, ii
        for ii, bit in enumerate(bitstring):
            # Only non zero entries represent a path edge
            if bit == '1':
                
                #Each bitstring has length n-1 where n is len(slices). The following statements index the list properly since
                # the ith edge is not represented in the string.
                if ii < i:
                    edge2 = ii
                elif ii >= i:
                    edge2 = ii+1
                path.append([edge1, edge2])
                
    return path

def get_slices(entire_bitstring):
    if isinstance(entire_bitstring, list):
        entire_bitstring = convert_to_bitstring(entire_bitstring)
        
    x = len(entire_bitstring)
    # Solve the quadratic equation to find n
    roots = np.roots([1, -1, -x])
    # n is the number of nodes
    n = int(max(roots))
    
    # Slice the bitstring into n slices, each of length n-1
    slices = []
    for i in range(n):
        # Each slice should be of length n-1
        temp = entire_bitstring[i * (n-1): (i+1) * (n-1)]
        slices.append(temp)
    return slices

def convert_to_bitstring(list_bitstring):
    bitstring = ''
    for i in list_bitstring:
        i = int(i)
        bitstring+=str(i)
        
    return bitstring

def primitive_runner(primitive, ansatz, hamiltonian=None, optimizer=None, optimal_parameters=None, shots = 10000):
    result = None
    
    # Run VQE if the primitive is an Estimator
    if isinstance(primitive, PrimEstimator):
        vqe = VQE(primitive, ansatz, optimizer)
        vqe_result = vqe.compute_minimum_eigenvalue(hamiltonian)
        return  vqe_result
    
    # Run Sampler if the primitive is a Sampler
    elif isinstance(primitive, PrimSampler):
        ansatz.measure_all()
        pub = [(ansatz, optimal_parameters)]
        job = primitive.run(pub, shots=shots)
        return job

    raise ValueError("Unsupported primitive type. Expected Estimator or Sampler.")
   ###################################################################


# Generates an n x n matrix that represents a graph of the VRP
n = 3
K = 1
instance = np.array([[0.0,         0.30233254, 0.26022551],
 [0.30233254, 0.0,         0.86583267],
 [0.26022551, 0.86583267, 0.0,        ]]
)
vrp = VehicleRouting(instance)

vrp_qp =vrp.to_quadratic_program()
quadratic_to_qubo = QuadraticProgramToQubo()
vrp_qubo = quadratic_to_qubo.convert(problem = vrp_qp)
ising_hamiltonian = vrp_qubo.to_ising()[0]

estimator = PrimEstimator()
sampler = PrimSampler()

ansatz = RealAmplitudes(num_qubits = ising_hamiltonian.num_qubits, reps = 2)
optimizer = COBYLA(maxiter = 5000)

estimator_results = primitive_runner(primitive = estimator,
                                     hamiltonian = ising_hamiltonian,
                                     ansatz = ansatz,
                                     optimizer = optimizer)

parameters = estimator_results.optimal_parameters
parameters = list(parameters.values())
shots_for_sampler = 10000
job = primitive_runner(primitive=sampler, ansatz=ansatz, optimal_parameters=parameters, shots=shots_for_sampler)

counts_dict = job.result()[0].data.meas.get_counts()

max_counts = (max(counts_dict.values()))
bitstring = [key for key, value in counts_dict.items() if value == max_counts][0]
quantum_path = format_solution(bitstring)

print(quantum_path)
```

### What should happen?

The quantum solution should match the classical CPLEX solution. Often times the correct solution is not sampled even once by the sampler.

The CPLEX solution to the matrix in the code above is [[0, 2], [1, 0], [2, 1]] which represents 0-2-1-0.

**Comment by t-imamichi** [comment]
Because this quantum algorithm is heuristic, the solution is not always optimal.

**Comment by RedWyrm13** [comment]
Hello!

Thanks for the response.

I have tried implementing this same thing using the QAOA and VQE algorithm classes and with the QAOAAnatz class. All of them return the same, incorrect solution. The implementation from the VRP tutorial shows that the quantum optimizer, which implements SamplingVQE, returns the same answer as CPLEX.

I would expect that the my results would match the CPLEX and SamplingVQE solutions for such a small graph of 3 nodes.

How can I know if it is just the heuristic nature of these algorithms that is yielding incorrect results or if my implementation is wrong? 

Changing the reps in the circuit, as well as the maxiter for the optimizer has not changed anything.

I tried increasing the penalty enforcement constant from the QuadraticToQubo class and that did not help either.

Thanks so much.
